src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,378 @@
+                {
+                    if (symbol is INamespaceSymbol namespaceSymbol)
+                    {
+                        containingNamespace = ConcatNamespace(containingNamespace, namespaceSymbol.Name);
+                        shouldVisitTypesInCurrentNamespace = !namespacesInScope.Contains(containingNamespace);
+
+                        foreach (var memberSymbol in namespaceSymbol.GetMembers())
+                        {
+                            VisitSymbol(memberSymbol, containingNamespace, shouldVisitTypesInCurrentNamespace);
+                        }
+                    }
+                    else if (shouldVisitTypesInCurrentNamespace
+                        && symbol is INamedTypeSymbol typeSymbol
+                        && typeSymbol.DeclaredAccessibility >= minimumAccessibility)
+                    {
+                        Debug.Assert(containingNamespace != null);
+                        var item = TypeImportCompletionItem.Create(typeSymbol, containingNamespace);
+                        builder.Add(item);
+                    }
+                }
+            }
+        }
+
+        private static string ConcatNamespace(string containingNamespace, string name)
+        {
+            Debug.Assert(name != null);
+            if (string.IsNullOrEmpty(containingNamespace))
+            {
+                return name;
+            }
+
+            _debug_total_namespace_concat++;
+            return containingNamespace + "." + name;
+        }
+
+        private string DebugText => string.Format(
+            DebugTextFormat,
+            _debug_total_compilation, _debug_total_compilation_decl, _debug_total_compilation_time,
+            _debug_total_pe, _debug_total_pe_decl, _debug_total_pe_time,
+            _debug_total_namespace_concat,
+            _debug_total_time_with_ItemCreation);
+
+        private int _debug_total_compilation = 0;
+        private int _debug_total_compilation_decl = 0;
+        private int _debug_total_compilation_time = 0;
+
+        private int _debug_total_pe = 0;
+        private int _debug_total_pe_decl = 0;
+        private int _debug_total_pe_time = 0;
+
+        private int _debug_total_time_with_ItemCreation = 0;
+
+        private static int _debug_total_namespace_concat = 0;
+
+        private void DebugClear()
+        {
+            _debug_total_compilation = 0;
+            _debug_total_compilation_decl = 0;
+            _debug_total_compilation_time = 0;
+
+            _debug_total_pe = 0;
+            _debug_total_pe_decl = 0;
+            _debug_total_pe_time = 0;
+
+            _debug_total_time_with_ItemCreation = 0;
+            _debug_total_namespace_concat = 0;
+        }
+
+        private const string DebugTextFormat = @"
+Total Compilations: {0}
+Total Declarations: {1}
+Elapsed time: {2}
+
+Total PEs: {3}
+Total Declarations: {4}
+Elapsed time: {5}
+
+Total namespace concat: {6}
+Total time: {7}";
+    }
+}
```

Comment:
i skipped this.  however, i was surprised by how much complexity there is.  was a lot of loic copy/pasted from some other provider?